README.md
# Algal Bloom Monthly Occurrence

Module responsable for extracting Algal Bloom Monthly Occurences in a region of interest based on images from Google Earth Engine API



### Dependencies

- Python 3.7.7 64-bit ou superior
- Modules: oauth2client earthengine-api matplotlib pandas numpy requests pillow natsort geojson argparse logging hashlib joblib gctraceback warnings



### Attention, before running:

Before running the script and after installing the libraries, you must authenticate with the Google Earth Engine API using the following command:

```
earthengine authenticate
```



### How to execute the default script?

python /path/to/abmo/script.py --lat_lon=-48.84725671390528,-22.04547298853004,-47.71712046185493,-23.21347463046867 --name=bbhr --date_start=1985-01-01 --date_end=2018-12-31




### What are the results?

The script will generate annual results of occurrence of algae and clouds blooming in the inserted study area. Therefore, a folder located in 'data' is created and named based on the date and version of the script executed. Example: /path/to/abmo/data/20200701_111426[v=V1-bbhr,date_start=1985-01-01,date_end=2001-12-31]. 

ATTENTION: big date range tends to lead to memory leak, stopping the script execution. It is always a good pratice to split the dates in two or three parts, unless you have a big amount of memory in your computer.

The following results are generated:

- timeseries.csv (Monthly time series of pixels, year, latitude, longitude and occurrences of algae bloom (based on the threshold of the Slope index) and clouds)
- occurrences.png (Graphs of occurrences separated annually from algal blooms)
- occurrences_clouds.png (Graphs of occurrences separated annually from clouds)
- occurrences.json (GeoJSON of occurrences by year and month with parameters (occurrence, not_occurrence, pct_occurrence, cloud, pct_cloud, year, month and instants) that can be imported into QGIS and filtered)
- tiff/{date}.zip (ZIP of GeoTIFFs comprehending bands Cloud/Only Water Body, Occurrence/Only Water Body and Not Occurrence/Only Water Body of the study area. If the script can not save those images in locally, will send them to Google Drive)


### Exporting GeoTIFFs to Google Drive

When using the 'save_collection_tiff' function, if the script can not save images locally, will send them to Google Drive to a folder called 'abmo_name_version.tiff' for user who is authenticated. Daily images used in the composition of the monthly time series are saved. These images will be separated by the following bands: Red, Green, Blue, Cloud/Only Water Body, Occurrence/Only Water Body and Not Occurrence/Only Water Body. However, after running the Abmo script, images are likely to take a while to be inserted into Drive due to processing time. It is necessary to wait approximately 1 day until they are all available, depending on the size of the study area.



### Example

```
# Import
import ee
from modules import abmo

# Initialize Google Earth Engine
ee.Initialize()

# folder where to save results
folder = "/path/to/desired/folder"

# create algorithm object
abmo = abmo.Abmo(lat_lon="-48.84725671390528,-22.04547298853004,-47.71712046185493,-23.21347463046867",
                date_start=dt.strptime("1985-01-01", "%Y-%m-%d"),
                date_end=dt.strptime("2018-12-31", "%Y-%m-%d"),
                date_start2=None,
                date_end2=None,
                sensor="landsat578",
                cache_path=folder, 
                force_cache=False,
                indice="mndwi,ndvi,fai,sabi,slope",
                min_occurrence=4,
                shapefile="users/pedroananias/bruno/bbhr")

# creating yearly timeseries
abmo.process_timeseries_data()

# save timeseries in csv file
abmo.save_dataset(df=abmo.df_timeseries, path=folder+'/timeseries.csv')

# save ocurrences plot (monthly and clouds)
abmo.save_occurrences_plot(df=abmo.df_timeseries, folder=folder)

# save geojson occurrences
abmo.save_occurrences_geojson(df=abmo.df_timeseries, folder+"/geojson")

# save daily images to Google Drive
abmo.save_collection_tiff(folder=folder+"/tiff", folderName="bbhr")
```

CHANGELOG
#########################################################################################################################################
# ### ABMO - Algal Bloom Monthly Occurences
# ### Script responsible for executing Algal Bloom Monthly Occurences in a region of interest
# ### Python 3.7 64Bits is required!
#########################################################################################################################################

### Change History
Version 1: 
- Repository creation

Version 2: 
- Fix nan values when generating geojson files

Version 3: 
- Fixes in the batch execution

Version 4:
- Added shapefile clipping for region of interest
- Added second date period for splitted seasons

batch.sh
#!/bin/bash 

# THIS DIR
BASEDIR="$( cd "$( dirname "$0" )" && pwd )"

# ARGUMENTS GENERAL
PYTHON=${1:-"python"}
SCRIPT="script.py"
CLEAR="sudo pkill -f /home/pedro/anaconda3"

# ATTRIBUTES
declare -a MIN_OCCS=(3 4)

# SHOW BASE DIR
echo "$PYTHON $BASEDIR/$SCRIPT"


############################################################################################
## PERIOD 1
LAT_LON="-48.56620427006758,-22.457495449468666,-47.9777042099919,-22.80261692472655"

# EXECUTIONS
for year in {1985..2018}
do
	for min_occ in "${MIN_OCCS[@]}"
	do
		eval "$PYTHON $BASEDIR/$SCRIPT --lat_lon=$LAT_LON --date_start=$year-01-01 --date_end=$year-12-31 --min_occurrence=$min_occ --seasonal --shapefile=users/pedroananias/bruno/bbhr"
	done
done
############################################################################################

modules/gee.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

##################################################################################################################
# ### Google Earth Engine API Support Functions
# ### Module responsible for storing Google Earth Engine's processing functions such as pixel extraction, etc.
##################################################################################################################

# Dependents Modules
import ee
import numpy as np
import traceback
import sys
from datetime import datetime as dt
from datetime import timedelta as td

# Configurations
# Defines the type of dummy value used
dummy               = -99999

# Define the indice that it will be used
indice_selected     = "mndwi,ndvi,sabi,fai,slope"
indice_thresholds   = {'mndwi': 0.0, 'ndvi': -0.15, 'sabi': -0.10, 'fai': -0.004, 'slope': -0.05}
min_occurrence      = 5

# Return the parameters of each sensor
def get_sensor_params(sensor: str):

  # COPERNICUS/S2_SR
  if sensor == "sentinel":
    return {
      'name': 'Sentinel-2/MSI [GEE Dataset = COPERNICUS/S2_SR]',
      'sensor': 'sentinel',
      'blue': 'B2',
      'red': 'B4',
      'green': 'B3',
      'swir': 'B11',
      'nir': 'B8',
      'c_nir': 834.05,
      'c_red': 664.75,
      'c_swir': 1612.05,
      'scale': 20,
      'start': dt.strptime("2017-03-28", "%Y-%m-%d"),
      'property_id': 'PRODUCT_ID'
    }

  # Modis MOD09GA.006
  elif sensor == "modis":
    return {
      'name': 'Modis MOD09GA.006 [GEE Dataset = MODIS/006/MOD09GA]',
      'sensor': 'modis',
      'blue': 'sur_refl_b03',
      'red': 'sur_refl_b01',
      'green': 'sur_refl_b04',
      'swir': 'sur_refl_b05',
      'nir': 'sur_refl_b02',
      'c_nir': 858.5,
      'c_red': 645.0,
      'c_swir': 1240.0,
      'scale': 500,
      'start': dt.strptime("2000-02-24", "%Y-%m-%d"),
      'property_id': 'system:id'
    }

  # Landsat-5/ETM
  elif sensor == "landsat5":
    return {
      'name': 'Landsat-5/ETM [GEE Dataset = LANDSAT/LT05/C01/T1_SR]',
      'sensor': 'landsat5',
      'blue': 'B1',
      'red': 'B3',
      'green': 'B2',
      'swir': 'B5',
      'nir': 'B4',
      'c_nir': 835.0,
      'c_red': 660.0,
      'c_swir': 1650.0,
      'scale': 30,
      'start': dt.strptime("1984-03-01", "%Y-%m-%d"),
      'end': dt.strptime("2011-10-28", "%Y-%m-%d"),
      'property_id': 'system:id'
    }

  # Landsat-7/ETM+
  elif sensor == "landsat7":
    return {
      'name': 'Landsat-7/ETM+ [GEE Dataset = LANDSAT/LE07/C01/T1_SR]',
      'sensor': 'landsat7',
      'blue': 'B1',
      'red': 'B3',
      'green': 'B2',
      'swir': 'B5',
      'nir': 'B4',
      'c_nir': 835.0,
      'c_red': 660.0,
      'c_swir': 1650.0,
      'scale': 30,
      'start': dt.strptime("1999-04-15", "%Y-%m-%d"),
      'property_id': 'system:id'
    }

  # Landsat-5/7/8 Merge
  elif sensor == "landsat578":
    return {
      'name': 'Landsat-5/7/8 Merge [GEE Dataset = LANDSAT/LT05/C01/T1_SR + LANDSAT/LE07/C01/T1_SR + LANDSAT/LC08/C01/T1_SR]',
      'sensor': 'landsat578',
      'blue': 'B2',
      'red': 'B4',
      'green': 'B3',
      'swir': 'B6',
      'nir': 'B5',
      'c_nir': 865,
      'c_red': 654.5,
      'c_swir': 1608.5,
      'scale': 30,
      'start': dt.strptime("1984-03-01", "%Y-%m-%d"),
      'property_id': 'system:id'
    }

  # Landsat-8/OLI
  else:
    return {
      'name': 'Landsat-8/OLI [GEE Dataset = LANDSAT/LC08/C01/T1_SR]',
      'sensor': 'landsat',
      'blue': 'B2',
      'red': 'B4',
      'green': 'B3',
      'swir': 'B6',
      'nir': 'B5',
      'c_nir': 865,
      'c_red': 654.5,
      'c_swir': 1608.5,
      'scale': 30,
      'start': dt.strptime("2013-02-11", "%Y-%m-%d"),
      'property_id': 'system:id'
    }


# Extract the coordinates of an image, pixel by pixel
def extract_latitude_longitude_pixel(image: ee.Image, geometry: ee.Geometry, bands: list, scale: int = 30, tile_scale: int = 16):

  # extract pixels lat and lons
  coordinates   = image.addBands(ee.Image.pixelLonLat()).select(['longitude', 'latitude']+bands)
  coordinates   = coordinates.reduceRegion(reducer=ee.Reducer.toList(), geometry=geometry, scale=scale, bestEffort=True, tileScale=tile_scale)

  # add bands
  band_values = []
  for band in bands:
    band_values.append(np.array(ee.List(coordinates.get(band)).getInfo(), dtype=np.float64))

  # build results
  band_values = np.array(band_values)
  result      = np.zeros(shape=(2+band_values.shape[0], band_values.shape[1]))
  result[0]   = np.array(ee.List(coordinates.get('longitude')).getInfo(), dtype=np.float64)
  result[1]   = np.array(ee.List(coordinates.get('latitude')).getInfo(), dtype=np.float64)
  for i, band_value in enumerate(band_values):
    result[i+2] = band_value

  # result
  return np.stack(result, axis=1)
  

# Return a collection for a given sensor
def get_sensor_collections(geometry: ee.Geometry, sensor: str = "landsat", dates: list = None):

  # COPERNICUS/S2_SR
  if sensor == "sentinel":
    collection          = ee.ImageCollection('COPERNICUS/S2_SR').filterBounds(geometry).filter(ee.Filter.lte('NODATA_PIXEL_PERCENTAGE','less_than', 100)).sort('system:time_start', True).map(apply_masks_sentinel)
    collection_water    = ee.ImageCollection('COPERNICUS/S2_SR').filterBounds(geometry).filterMetadata('CLOUDY_PIXEL_PERCENTAGE','less_than', 10).filter(ee.Filter.lte('NODATA_PIXEL_PERCENTAGE',0)).sort('system:time_start', True).map(apply_masks_sentinel)

  # Modis MOD09GA.006
  elif sensor == "modis":
    collection          = ee.ImageCollection('MODIS/006/MOD09GA').filterBounds(geometry).sort('system:time_start', True).map(apply_waterleaving_reflectance_correction_modis).map(apply_masks_modis)
    collection_water    = ee.ImageCollection('MODIS/006/MOD44W').filterBounds(geometry)

  # Landsat-5/ETM
  elif sensor == "landsat5":
    collection          = ee.ImageCollection('LANDSAT/LT05/C01/T1_SR').filterBounds(geometry).sort('system:time_start', True).map(apply_masks_landsat5)  
    collection_water    = ee.ImageCollection('GLCF/GLS_WATER').filterBounds(geometry)

  # Landsat-7/ETM+
  elif sensor == "landsat7":
    collection          = ee.ImageCollection('LANDSAT/LE07/C01/T1_SR').filterBounds(geometry).sort('system:time_start', True).map(apply_masks_landsat7)  
    collection_water    = ee.ImageCollection('GLCF/GLS_WATER').filterBounds(geometry)

  # Landsat-5/7/8 - Merge
  elif sensor == "landsat578":
    
    # Landsat-5/ETM
    collection5         = ee.ImageCollection('LANDSAT/LT05/C01/T1_SR').filterDate(get_sensor_params("landsat5")["start"],get_sensor_params("landsat5")["end"]).filterBounds(geometry).sort('system:time_start', True).select(["B1","B2","B3","B4","B5","B6","pixel_qa"]).map(apply_masks_landsat5)  
    
    # Landsat-7/ETM+
    collection7         = ee.ImageCollection('LANDSAT/LE07/C01/T1_SR').filterDate(get_sensor_params("landsat5")["end"],get_sensor_params("landsat")["start"]).filterBounds(geometry).sort('system:time_start', True).select(["B1","B2","B3","B4","B5","B6","pixel_qa"]).map(apply_masks_landsat7)  
    
    # Landsat-8/OLI
    collection8         = ee.ImageCollection('LANDSAT/LC08/C01/T1_SR').filterBounds(geometry).sort('system:time_start', True).select(["B1","B2","B3","B4","B5","B6","pixel_qa"]).map(apply_masks_landsat)  
    
    # Landsat Merge
    collection          = collection8.merge(collection7.merge(collection5)).sort('system:time_start', True)
    collection_water    = ee.ImageCollection('GLCF/GLS_WATER').filterBounds(geometry)

  # Landsat-8/OLI
  else:
    collection          = ee.ImageCollection('LANDSAT/LC08/C01/T1_SR').filterBounds(geometry).sort('system:time_start', True).map(apply_masks_landsat)  
    collection_water    = ee.ImageCollection('GLCF/GLS_WATER').filterBounds(geometry)
    
  # Filter dates
  if dates:
    collection = collection.filterDate(dates[0],dates[1])
    if not collection_water:
      collection_water = collection.filterDate(dt.strftime(dt.strptime(dates[0], "%Y-%m-%d")-td(days=180), "%Y-%m-%d"), dt.strftime(dt.strptime(dates[1], "%Y-%m-%d")+td(days=180), "%Y-%m-%d")).sort('system:time_start', True)

  return collection, collection_water


# Cloud Mask and Cloud Shadows
def mask_cloud_shadow(image, sensor: str):

  # COPERNICUS/S2_SR
  # Values: https://earth.esa.int/web/sentinel/technical-guides/sentinel-2-msi/level-2a/algorithm
  if sensor == "sentinel":
      qa = image.select('QA60')
      return qa.bitwiseAnd(1 << 10).eq(0) and (qa.bitwiseAnd(1 << 11).eq(0))

  # Modis MOD09GA.006
  # Values: https://developers.google.com/earth-engine/datasets/catalog/MODIS_006_MOD09GA
  elif sensor == "modis":
      qa = image.select('state_1km')
      return qa.bitwiseAnd(1 << 10).eq(0)

  # Landsat-5/ETM
  # Binary Values: https://landsat.usgs.gov/sites/default/files/documents/landsat_QA_tools_userguide.pdf
  # Absolute Values: https://prd-wret.s3.us-west-2.amazonaws.com/assets/palladium/production/atoms/files/LSDS-1370_L4-7_SurfaceReflectance-LEDAPS_ProductGuide-v2.pdf
  elif sensor == "landsat5":
      qa = image.select('pixel_qa')
      return (qa.bitwiseAnd(1 << 5).eq(0) and (qa.bitwiseAnd(1 << 7).eq(0))) or (qa.bitwiseAnd(1 << 3).eq(0))

  # Landsat-7/ETM+
  # Binary Values: https://landsat.usgs.gov/sites/default/files/documents/landsat_QA_tools_userguide.pdf
  # Absolute Values: https://prd-wret.s3.us-west-2.amazonaws.com/assets/palladium/production/atoms/files/LSDS-1370_L4-7_SurfaceReflectance-LEDAPS_ProductGuide-v2.pdf
  elif sensor == "landsat7":
      qa = image.select('pixel_qa')
      return (qa.bitwiseAnd(1 << 5).eq(0) and (qa.bitwiseAnd(1 << 7).eq(0))) or (qa.bitwiseAnd(1 << 3).eq(0))

  # Landsat-8/OLI
  # Binary Values: https://landsat.usgs.gov/sites/default/files/documents/landsat_QA_tools_userguide.pdf
  # Absolute Values: https://prd-wret.s3.us-west-2.amazonaws.com/assets/palladium/production/atoms/files/LSDS-1368_L8_SurfaceReflectanceCode-LASRC_ProductGuide-v2.pdf
  else:
      qa = image.select('pixel_qa')
      return qa.bitwiseAnd(1 << 3).eq(0) and (qa.bitwiseAnd(1 << 5).eq(0))


# Water Mask
def mask_water(image, sensor: str):

  # Sentinel-2/MSI
  # Values: https://earth.esa.int/web/sentinel/technical-guides/sentinel-2-msi/level-2a/algorithm
  if sensor == "sentinel":
      qa = image.select('SCL')
      return qa.eq(6).Not()

  # Modis MOD09GA.006
  # Values: https://developers.google.com/earth-engine/datasets/catalog/MODIS_006_MOD09GA
  elif sensor == "modis":
      qa = image.select('state_1km')
      return ((qa.bitwiseAnd(1 << 3).neq(0)) and (qa.bitwiseAnd(1 << 4).eq(0) or qa.bitwiseAnd(1 << 5).eq(0)))

  # Landsat-5/ETM
  # Binary Values: https://landsat.usgs.gov/sites/default/files/documents/landsat_QA_tools_userguide.pdf
  # Absolute Values: https://prd-wret.s3.us-west-2.amazonaws.com/assets/palladium/production/atoms/files/LSDS-1370_L4-7_SurfaceReflectance-LEDAPS_ProductGuide-v2.pdf
  elif sensor == "landsat5":
      qa = image.select('pixel_qa')
      return qa.bitwiseAnd(1 << 2).eq(0)

  # Landsat-7/ETM+
  # Binary Values: https://landsat.usgs.gov/sites/default/files/documents/landsat_QA_tools_userguide.pdf
  # Absolute Values: https://prd-wret.s3.us-west-2.amazonaws.com/assets/palladium/production/atoms/files/LSDS-1370_L4-7_SurfaceReflectance-LEDAPS_ProductGuide-v2.pdf
  elif sensor == "landsat7":
      qa = image.select('pixel_qa')
      return qa.bitwiseAnd(1 << 2).eq(0)
      
  # Landsat-8/OLI
  # Binary Values: https://landsat.usgs.gov/sites/default/files/documents/landsat_QA_tools_userguide.pdf
  # Absolute Values: https://prd-wret.s3.us-west-2.amazonaws.com/assets/palladium/production/atoms/files/LSDS-1368_L8_SurfaceReflectanceCode-LASRC_ProductGuide-v2.pdf
  else:
      qa = image.select('pixel_qa')
      return qa.bitwiseAnd(1 << 2).eq(0)


# Sentinel-2/MSI
def apply_masks_sentinel(image):
  return apply_masks(image, get_sensor_params("sentinel"))

# Landsat-5/ETM
def apply_masks_landsat5(image):
  return apply_masks(image, get_sensor_params("landsat5"))

# Landsat-7/ETM+
def apply_masks_landsat7(image):
  return apply_masks(image, get_sensor_params("landsat7"))

# Lantsat-8/OLI
def apply_masks_landsat(image):
  return apply_masks(image, get_sensor_params("landsat"))

# Modis MOD09GA.006
def apply_masks_modis(image):
  return apply_masks(image, get_sensor_params("modis"))


# Modis MOD09GA.006 Water-leaving reflectance correction
# Rc(i) = R(i) - min(R(NIR):R(SWIR)) (Jia et al. 2019)
def apply_waterleaving_reflectance_correction_modis(image):

  # select bands (NIR:SWIR)
  swir1 = image.select('sur_refl_b05')
  swir2 = image.select('sur_refl_b06')
  swir3 = image.select('sur_refl_b07')

  # compute min
  min_swir1_swir2 = swir1.min(swir2)
  wl_re_corr = min_swir1_swir2.min(swir3).rename('wl_re_corr')

  # apply correction RED, NIR and SWIR
  red     = image.expression('red - wl_re_corr', { 'red': image.select('sur_refl_b01'), 'wl_re_corr': wl_re_corr.select('wl_re_corr')}).rename('sur_refl_b01')
  green   = image.expression('green - wl_re_corr', { 'green': image.select('sur_refl_b04'), 'wl_re_corr': wl_re_corr.select('wl_re_corr')}).rename('sur_refl_b04')
  blue    = image.expression('blue - wl_re_corr', { 'blue': image.select('sur_refl_b03'), 'wl_re_corr': wl_re_corr.select('wl_re_corr')}).rename('sur_refl_b03')
  nir     = image.expression('nir - wl_re_corr', { 'nir': image.select('sur_refl_b02'), 'wl_re_corr': wl_re_corr.select('wl_re_corr')}).rename('sur_refl_b02')
  swir    = image.expression('swir - wl_re_corr', { 'swir': image.select('sur_refl_b05'), 'wl_re_corr': wl_re_corr.select('wl_re_corr')}).rename('sur_refl_b05')

  # add new band
  return image.addBands([red,green,blue,nir,swir], overwrite=True)


# Apply indexes and apply water masks, cloud
def apply_masks(image, params: dict):

  # Apply the water and cloud masks and cloud shade
  blank           = ee.Image(abs(dummy))
  water           = blank.updateMask(mask_water(image, params['sensor']).Not()).rename('water')
  cloud           = ee.Image(1).updateMask(mask_cloud_shadow(image, params['sensor']).Not()).rename('cloud')
  nocloud         = blank.updateMask(mask_cloud_shadow(image, params['sensor'])).rename('nocloud')
  water_nocloud   = water.updateMask(nocloud).rename('water_nocloud')

  # conditions
  array_condition = []
  for i in indice_selected.split(","):
    operator = ">" if not i == "mndwi" else "<"
    array_condition.append("("+i+" "+operator+" "+str(indice_thresholds[i])+")")
  condition_index = " + ".join(array_condition)
      
  # Apply the indexes available in the image
  mndwi           = image.expression('(green - swir) / (green + swir)',{'swir':image.select(params['swir']).multiply(0.0001),'green':image.select(params['green']).multiply(0.0001)}).rename('mndwi').cast({"mndwi": "double"}) # Xu (2006)
  ndvi            = image.expression('(nir - red) / (nir + red)',{'nir':image.select(params['nir']).multiply(0.0001),'red':image.select(params['red']).multiply(0.0001)}).rename('ndvi').cast({"ndvi": "double"})
  fai             = image.expression('nir - (red + (swir - red) * ((c_nir - c_red) / (c_swir - c_red)))',{'swir':image.select(params['swir']).multiply(0.0001),'nir':image.select(params['nir']).multiply(0.0001),'red':image.select(params['red']).multiply(0.0001),'c_nir':params['c_nir'],'c_red':params['c_red'],'c_swir':params['c_swir']}).rename('fai').cast({"fai": "double"}) # Oyama et al (2015)
  slope           = image.expression('((red - nir) / (c_red - c_nir)) * 1000',{'nir':image.select(params['nir']).multiply(0.0001),'red':image.select(params['red']).multiply(0.0001),'c_nir':params['c_nir'],'c_red':params['c_red']}).rename('slope').cast({"slope": "double"}) # Ogashawara, Li, and Moreno-Madriñán (2017)
  sabi            = image.expression('(nir - red) / (blue + green)',{'nir':image.select(params['nir']).multiply(0.0001),'red':image.select(params['red']).multiply(0.0001),'blue':image.select(params['blue']).multiply(0.0001),'green':image.select(params['green']).multiply(0.0001)}).rename('sabi').cast({"sabi": "double"}) # Alawadi (2010)
  
  # all indices agree
  label           = image.expression('(cloud == 1) ? -1 : '+condition_index, {'ndvi': ndvi.select('ndvi'), 'fai': fai.select('fai'), 'slope': slope.select('slope'), 'sabi': sabi.select('sabi'), 'mndwi': mndwi.select('mndwi'), 'cloud': cloud.select('cloud')}).rename('label')

  # occurrences
  occurrence      = label.expression('(label >= '+str(min_occurrence)+') ? 1 : 0', {'label': label.select('label')}).rename('occurrence')
  not_occurrence  = label.expression('(label < '+str(min_occurrence)+') ? 1 : 0', {'label': label.select('label')}).rename('not_occurrence')

  # Create the bands to the image and return it
  return image.addBands([water, water_nocloud, cloud, nocloud, mndwi, ndvi, fai, sabi, slope, label, occurrence, not_occurrence])


# Apply to a mask and return image with the new band
def apply_mask(image, mask, band_from, band_to, remove_empty_pixels = False):
  image_mask = image.select(band_from).updateMask(mask).rename(band_to)
  if remove_empty_pixels:
    return image.addBands([image_mask])
  image_mask = ee.Image(abs(dummy)).blend(image_mask).rename(band_to)
  return image.addBands([image_mask])


# Get image min and max values
def get_image_min_max(image: ee.Image, geometry: ee.Geometry, scale: int = None, tile_scale: int = 16):
  if scale is None:
    scale = image.projection().nominalScale()
  image = image.addBands(ee.Image.pixelLonLat())
  return image.reduceRegion(reducer=ee.Reducer.min(), geometry=geometry, scale=scale, bestEffort=True, tileScale=tile_scale).getInfo(), image.reduceRegion(reducer=ee.Reducer.max(), geometry=geometry, scale=scale, bestEffort=True, tileScale=tile_scale).getInfo()


# Get image counters
def get_image_counters(image: ee.Image, geometry: ee.Geometry, scale: int = None, tile_scale: int = 16):
  if scale is None:
    scale = image.projection().nominalScale()
  return image.reduceRegion(reducer=ee.Reducer.count(), geometry=geometry, scale=scale, bestEffort=True, tileScale=tile_scale).getInfo()


# Get geometry from diagonal points
def get_geometry_from_lat_lon(lat_lon: str):

  # Selection of coordinates (colon, lat and lon separated by comma, all together) and dates by the user (two dates, beginning and end, separated by commas)
  x1,y1,x2,y2 = lat_lon.split(",")

  # Assemble Geometry on Google Earth Engine
  geometry = ee.Geometry.Polygon(
        [[[float(x1),float(y2)],
          [float(x2),float(y2)],
          [float(x2),float(y1)],
          [float(x1),float(y1)],
          [float(x1),float(y2)]]])

  # return the geometry
  return geometry

generate_geojson.py
import geojson
import pandas as pd
import glob
import os

# read the results files in data directory
files = glob.glob("data/results*.csv")

# go through all read files
for file in files:
  df = pd.read_csv(r""+file)
  features = []
  for index, row in df.iterrows():
      features.append(geojson.Feature(geometry=geojson.Point((row['lat'], row['lon'])), properties={"occurrence": int(row['occurrence']), "not_occurrence": int(row['not_occurrence']), "pct_occurrence": int(row['pct_occurrence']), "cloud": int(row['cloud']), "pct_cloud": int(row['pct_cloud']), "year": int(row['year']), "season": str(row['season']), "instants": int(row['instants'])}))
  fc = geojson.FeatureCollection(features)
  f = open(os.path.splitext(file)[0].replace("results", "occurrences")+".json","w")
  geojson.dump(fc, f)
  f.close()

script.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

#########################################################################################################################################
# ### ABMO - Algal Bloom Monthly Occurences
# ### Script responsible for executing Algal Bloom Monthly Occurences in a region of interest
# ### Python 3.7 64Bits is required!
#########################################################################################################################################

# ### Version
version = "V4"


# ### Module imports

# Main
import ee
import time
import os
import sys
import argparse
import traceback
import pandas as pd

# Sub
from datetime import datetime as dt

# Extras modules
from modules import misc, gee, abmo


# ### Script args parsing

# starting arg parser
parser = argparse.ArgumentParser(description=version)

# create arguments
parser.add_argument('--lat_lon', dest='lat_lon', action='store', default="-48.56620427006758,-22.457495449468666,-47.9777042099919,-22.80261692472655",
                   help="Two diagnal points (Latitude 1, Longitude 1, Latitude 2, Longitude 2) of the study area")
parser.add_argument('--date_start', dest='date_start', action='store', default="1985-01-01",
                   help="Date to start time series")
parser.add_argument('--date_end', dest='date_end', action='store', default="2001-12-31",
                   help="Date to end time series")
parser.add_argument('--date_start2', dest='date_start2', action='store', default=None,
                   help="Date to start time series (second period)")
parser.add_argument('--date_end2', dest='date_end2', action='store', default=None,
                   help="Date to end time series (second period)")
parser.add_argument('--name', dest='name', action='store', default="bbhr",
                   help="Place where to save generated files")
parser.add_argument('--sensor', dest='sensor', action='store', default="landsat578",
                   help="Define which sensor will be used")
parser.add_argument('--indice', dest='indice', action='store', default="mndwi,ndvi,fai,sabi,slope",
                   help="Define which indice will be used to determine algal blooms (mndwi, ndvi, fai, sabi e/ou slope)")
parser.add_argument('--min_occurrence', dest='min_occurrence', type=int, action='store', default=4,
                   help="Define how many indices will have to match in order to determine pixel as algal bloom occurrence")
parser.add_argument('--seasonal', dest='seasonal', action='store_true',
                   help="Define if pixels will be reduced by using seasons instead of months")
parser.add_argument('--shapefile', dest='shapefile', action='store',
                   help="Use a shapefile to clip a region of interest")
parser.add_argument('--force_cache', dest='force_cache', action='store_true',
                   help="Force cache reseting to prevent image errors")

# parsing arguments
args = parser.parse_args()




# ### Start

try:

  # Start script time counter
  start_time = time.time()

  # Google Earth Engine API initialization
  ee.Initialize()



  # ### Working directory

  # Data path
  folderRoot = os.path.dirname(os.path.realpath(__file__))+'/data'
  if not os.path.exists(folderRoot):
    os.mkdir(folderRoot)

  # Images path
  folderCache = os.path.dirname(os.path.realpath(__file__))+'/cache'
  if not os.path.exists(folderCache):
    os.mkdir(folderCache)


  
  # ### ABMO execution

  # folder to save results from algorithm at
  folder = folderRoot+'/'+dt.now().strftime("%Y%m%d_%H%M%S")+'[v='+str(version)+'-'+str(args.name)+',dstart='+str(args.date_start)+',dend='+str(args.date_end)+',dstart2='+str(args.date_start2)+',dend2='+str(args.date_end2)+',i='+str(args.indice)+',moc='+str(args.min_occurrence)+',s='+str(args.seasonal)+']'
  if not os.path.exists(folder):
    os.mkdir(folder)

  # create algorithm
  abmo = abmo.Abmo(lat_lon=args.lat_lon,
                   date_start=dt.strptime(args.date_start, "%Y-%m-%d"),
                   date_end=dt.strptime(args.date_end, "%Y-%m-%d"),
                   date_start2=dt.strptime(args.date_start2, "%Y-%m-%d") if not args.date_start2 is None else None,
                   date_end2=dt.strptime(args.date_end2, "%Y-%m-%d") if not args.date_end2 is None else None,
                   sensor=args.sensor,
                   cache_path=folderCache, 
                   force_cache=args.force_cache,
                   indice=args.indice,
                   min_occurrence=args.min_occurrence,
                   seasonal=args.seasonal,
                   shapefile=args.shapefile)

  # preprocessing
  abmo.process_timeseries_data()

  # create plot
  abmo.save_occurrences_plot(df=abmo.df_timeseries, folder=folder)

  # save timeseries in csv file
  abmo.save_dataset(df=abmo.df_timeseries, path=folder+'/timeseries[dstart='+str(args.date_start)+',dend='+str(args.date_end)+',moc='+str(args.min_occurrence)+',s='+str(args.seasonal)+'].csv')

  # save geojson occurrences and clouds
  abmo.save_occurrences_geojson(df=abmo.df_timeseries, path=folder+'/occurrences[dstart='+str(args.date_start)+',dend='+str(args.date_end)+',moc='+str(args.min_occurrence)+',s='+str(args.seasonal)+'].json')

  # save images to Local Folder (first try, based on image size) or to your Google Drive
  #abmo.save_collection_tiff(folder=folder+"/tiff", folderName=args.name+"_"+version, rgb=False).

  # results
  # add results and save it on disk
  abmo.df_timeseries = abmo.df_timeseries.drop(['label'], axis=1)
  path_df_timeseries = folderRoot+'/results[moc='+str(args.min_occurrence)+',s='+str(args.seasonal)+'].csv'
  df_timeseries = pd.read_csv(path_df_timeseries).drop(['Unnamed: 0'], axis=1, errors="ignore").append(abmo.df_timeseries) if os.path.exists(path_df_timeseries) else abmo.df_timeseries.copy(deep=True)
  df_timeseries.to_csv(r''+path_df_timeseries)

  # ### Script termination notice
  script_time_all = time.time() - start_time
  debug = "***** Script execution completed successfully (-- %s seconds --) *****" %(script_time_all)
  print()
  print(debug)

except:

    # ### Script execution error warning

    # Execution
    print()
    print()
    debug = "***** Error on script execution: "+str(traceback.format_exc())
    print(debug)

    # Removes the folder created initially with the result of execution
    script_time_all = time.time() - start_time
    debug = "***** Script execution could not be completed (-- %s seconds --) *****" %(script_time_all)
    print(debug)
